### PR TITLE
explicit media toggle and small bug fixes

### LINKED ISSUE #152: toggle async media loading
## Issue
With async toot loading and http streams I am running into issues when media is loaded asynchronously.

## Suggestion
Currently `mastodon-tl--show-avatars-p` is automatically toggled based on whether or not the client has `imagemagick`.  For debugging purposes, add a similar toggle `mastodon-tl--display-media-p` that:
1. prevents media stubs from being put into toots
2. prevents the async loading of those images
3. prevents loading avatars

### PR DESCRIPTION
**Pull request by alexjgriffith**
## Summary
This PR closes #152  and extends the fix for #150 to all places where ids were being converted from numbers to strings. It also adds some small fixes to the behaviour of threads and fixes boosting and favouriting (unfavouriting and unboosting still don't work).

## Changes
1. We now kill the http get request buffer once JSON has been extracted. 
2. `mastodon-tl--as-string` was implemented and replaced any occurrence of `number-to-string` or `int-to-string`
3. Added variable `mastodon-tl--display-media-p`. By default it is `'t` but can be made a local buffer variable and set to `nil`. When `nil` rather than displaying the media it just provides a link `Media::<link>`
4. Fixed checking for faves and boosts, they should now render properly. The return from `json-read-from-string` for `nil` is `:json-false` which evaluates to `'t` in elisp.
5. Fixed the `format` string that gets printed when faving and boosting
6. Fixed `mastodon-tl--thread` updating and requesting and changed its behaviour such that it tries to open the original toot thread rather than the boosted thread. 
7. Added tests for both the new `mastodon-tl--as-string` function and the `mastodon-tl--toot-id` utility.

### CODE REVIEW

**Comment by alexjgriffith** [comment]
Merged commits and replaced `alist-get` with `(cdr (assoc))`.

File: lisp/mastodon-http.el [hunk 1]
```diff
@@ -100,13 +100,14 @@ Pass response buffer to CALLBACK function."
 (defun mastodon-http--get-json (url)
   "Make GET request to URL. Return JSON response vector."
   (let ((json-vector
-         (with-current-buffer (mastodon-http--get url)
+         (with-current-bufer (mastodon-http--get url)
```
Comment (hdurer) [inline lisp/mastodon-http.el]: Surely that's wrong.  Should stay `buffer`

File: lisp/mastodon-http.el [hunk 2]
```diff
@@ -100,7 +100,7 @@ Pass response buffer to CALLBACK function."
 (defun mastodon-http--get-json (url)
   "Make GET request to URL. Return JSON response vector."
   (let ((json-vector
-         (with-current-bufer (mastodon-http--get url)
+         (with-current-buffer (mastodon-http--get url)
```
Comment (hdurer) [inline lisp/mastodon-http.el]: Why not squash this commit with the previous one to keep the history sane and readable?

File: lisp/mastodon-tl.el [hunk 3]
```diff
@@ -293,11 +293,14 @@ also render the html"
...
-                            (mastodon-media--get-media-link-rendering
-                             preview-url)))
-                        media-attachements "")))
-    (if (not (equal media-string ""))
-        (concat "\n" media-string ) "")))
+                            (if mastodon-tl--display-media-p
+                                (mastodon-media--get-media-link-rendering
+                                 preview-url)
+                              (concat "Media::" preview-url ))))
+                        media-attachements "\n")))
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: Why do we re-add the ugly newlines between images?  That's only needed if `mastodon-tl--display-media-p` is nil, or not?

File: lisp/mastodon-tl.el [hunk 4]
```diff
@@ -293,11 +293,14 @@ also render the html"
...
+    (if (not (equal media-string ""))
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: use `string=` and only do this newline prepending if not displaying media.

File: lisp/mastodon-tl.el [hunk 5]
```diff
@@ -317,8 +320,7 @@ also render the html"
     (mastodon-tl--spoiler toot)
     ;; remove two trailing newlines
     (substring (mastodon-tl--content toot) 0 -2)
-    (when mastodon-tl--display-media-p
-      (mastodon-tl--media toot))
+    (mastodon-tl--media toot)
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: We have only just made this change and now it is being undone again.  Would it make sense to just squash this commit with the first one?

File: lisp/mastodon-tl.el [hunk 6]
```diff
@@ -1,4 +1,4 @@
-;;; mastodon-tl.el --- HTTP request/response functions for mastodon.el  -*- lexical-binding: t -*-
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: a) all the files seem to use 3 semicolons, why change it in only one?
b) this seems unrelated to this PR.

File: lisp/mastodon-tl.el [hunk 7]
```diff
@@ -399,21 +399,35 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
   (goto-char (point-max))
   (mastodon-tl--property 'toot-id t))
 
+(defun mastodon-tl--as-string(id)
+  (cond ((numberp id )
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: Add doc string

File: lisp/mastodon-tl.el [hunk 8]
```diff
@@ -399,21 +399,35 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
   (goto-char (point-max))
   (mastodon-tl--property 'toot-id t))
 
+(defun mastodon-tl--as-string(id)
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: `mastodon-tl--id-as-string`?

File: lisp/mastodon-tl.el [hunk 9]
```diff
@@ -399,21 +399,35 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
...
-  (let* ((id (number-to-string (mastodon-tl--property 'toot-id)))
+  (let* ((id (mastodon-tl--as-string (mastodon-tl--property 'toot-id)))
          (url (mastodon-http--api (format "statuses/%s/context" id)))
          (buffer (format "*mastodon-thread-%s*" id))
          (toot (mastodon-tl--property 'toot-json))
          (context (mastodon-http--get-json url)))
     (with-output-to-temp-buffer buffer
       (switch-to-buffer buffer)
+      (mastodon-mode)
+      (setq mastodon-tl--buffer-spec
```
Comment (hdurer) [inline lisp/mastodon-tl.el:433]: Nit: This change seems unrelated to what the commit message says...

File: lisp/mastodon-toot.el [hunk 10]
```diff
@@ -99,7 +99,7 @@ Remove MARKER if RM is non-nil."
     (mastodon-toot--action action
                            (lambda ()
                              (mastodon-toot--action-success "F" remove)
-                             (message (format "%sd #%d" action id))))))
+                             (message (format "%s #%d" action id))))))
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: Now that ids are strings, wouldn't the second formatting param also be `%s`?

File: lisp/mastodon-tl.el [hunk 11]
```diff
@@ -408,10 +408,24 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
...
+  "Find approproiate toot id for toot under point.
+
+If the toot has been boosted use the id found in the
+reblog portion of the toot. Otherwise, use the body of
+the toot. This is the same behaviour as the mastodon.social
+webapp"
+  (let* ((json (mastodon-tl--property 'toot-json))
+         (id (alist-get 'id json ))
+         (reblog (alist-get 'reblog json )))
+    (if reblog
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: Should this be wrapped in the new "make it a string" helper function to ensure we return valid ids?

File: lisp/mastodon-tl.el [hunk 12]
```diff
@@ -423,7 +437,7 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
             `(buffer-name ,buffer
                           endpoint ,(format "statuses/%s/context" id)
                           update-function
-                          mastodon-tl--timeline))
+                          (lambda(toot) (message "END of thread."))))
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: Nit: space after `lambda`

File: lisp/mastodon-tl.el [hunk 13]
```diff
@@ -236,8 +235,8 @@ TIME-STAMP is assumed to be in the past."
   "Generate byline for TOOT."
   (let ((id (cdr (assoc 'id toot)))
         (parsed-time (date-to-time (mastodon-tl--field 'created_at toot)))
-        (faved (equal ":json-true" (mastodon-tl--field 'favourited toot)))
-        (boosted (equal ":json-true" (mastodon-tl--field 'reblogged toot))))
+        (faved (equal 't (mastodon-tl--field 'favourited toot)))
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: Whoops!  That looks like either our tests are wrong or some test is missing?

**Review by hdurer — COMMENTED** [review]
First pass through these changes.  I find it very hard to read and understand this PR.  I think it would be preferable if you squashed all this into 2 or 3 commits.

There is really no need to keep all the false starts, accidental changes and undoing of those accidental changes.

File: lisp/mastodon-tl.el (hunk 8 above)
Comment (alexjgriffith) [reply lisp/mastodon-tl.el]: This could be used anywhere we wish to convert a number to a string. To generalize I will rename the variable `ID` to `NUMERIC`.

File: lisp/mastodon-tl.el (hunk 9 above)
Comment (alexjgriffith) [reply lisp/mastodon-tl.el:433]: Updated the commit message to include a line addressing this move. I was having odd behaviour in the thread buffer if i defined mastodon-tl--buffer-spec` before initializing `mastodon-mode`.

**Comment by alexjgriffith** [comment]
Squashed the commits, for real this time, and generalized the variable names in `mastodon-tl--as-string`. This function should really be put in a utility module at some point.

File: lisp/mastodon-tl.el (hunk 13 above)
Comment (hdurer) [reply lisp/mastodon-tl.el]: Ah ok, I now get it (reading the commit message).

Resolved.

File: lisp/mastodon-tl.el (hunk 3 above)
Comment (hdurer) [reply lisp/mastodon-tl.el]: I still think we should not add newlines between images when we do show images...

File: lisp/mastodon-tl.el (hunk 8 above)
Comment (hdurer) [reply lisp/mastodon-tl.el]: ok.

**Review by hdurer — DISMISSED** [review]
Seems fine.  I suggest you still squash these two into one before merging into develop.

I still think adding newlines between images is wrong but I guess that's a matter of taste?

**Comment by alexjgriffith** [comment]
Thanks for the quick feedback! We certainly do not need newlines between images. Will make a fix there and squash the commits.

**Review by hdurer — APPROVED** [review]
LGTM
